### PR TITLE
Add docs for new Caddy image

### PR DESCRIPTION
**Pull request by hairyhenderson**
Docs to go with the new Caddy image: https://github.com/docker-library/official-images/pull/7574

Signed-off-by: Dave Henderson <dhenderson@gmail.com>

### CODE REVIEW

File: caddy/content.md [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+
+# What is Caddy?
+
+[Caddy 2](https://caddyserver.com/) is a powerful, enterprise-ready, open source web server with automatic HTTPS written in Go.
+
+## How to use this image
+
+#### ⚠️ A note about persisted data
+
+Caddy requires write access to two locations: a [data directory](https://caddyserver.com/docs/conventions#data-directory), and a [configuration directory](https://caddyserver.com/docs/conventions#configuration-directory). While it's not necessary to persist the files stored in the configuration directory, it can be convenient. However, it's very important to persist the data directory.
```
[the remark this answers is not in the document]
Comment (hairyhenderson) [reply caddy/content.md:11]: The original is more correct IMO - the data directory _itself_ (and its contents) need to be persisted. What was unclear about the original?

File: caddy/content.md [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+
+From the docs:
+
+> The data directory must not be treated like a cache. Its contents are not ephemeral or merely for the sake of performance. Caddy stores TLS certificates, private keys, OCSP staples, and other necessary information to the data directory. It should not be purged without an understanding of the implications.
```
[the remark this answers is not in the document]
Comment (mholt) [reply caddy/content.md:15]: This is a quoted section from our website, so I will update our website to comply with this change, which I think is fine - thanks!

File: caddy/README-short.txt [hunk 3]
```diff
@@ -0,0 +1 @@
+Caddy 2 is a powerful, enterprise-ready, open source web server with automatic HTTPS written in Go.
```
[the remark this answers is not in the document]
Comment (hairyhenderson) [reply caddy/README-short.txt:1]: I see both `open source` and `open-source` used in this repo - is there a reason this should be changed? I'm not aware of any formal convention.

Also this is the text directly from https://caddyserver.com/ so I'm a bit reluctant to change this without good reason.

File: caddy/content.md [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+### Basic Usage
+
+The default config file simply serves files from `/usr/share/caddy`, so if you want to serve `index.html` from the current working directory:
+
+```console
+$ echo "hello world" > index.html
+$ docker run -d -p 80:80 \
+    -v $(pwd)/index.html:/usr/share/caddy/index.html \
+    -v caddy_data:/data \
+    caddy/caddy
```
Comment (tianon) [inline caddy/content.md:32]:
```suggestion
    %%IMAGE%%
```

File: caddy/content.md [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+hello world
+```
+
+To override the default [`Caddyfile`](https://github.com/caddyserver/dist/blob/master/config/Caddyfile), you can mount a new one at `/etc/caddy/Caddyfile`:
+
+```console
+$ docker run -d -p 80:80 \
+    -v $(pwd)/Caddyfile:/etc/caddy/Caddyfile \
+    -v caddy_data:/data \
+    caddy/caddy
```
Comment (tianon) [inline caddy/content.md:44]:
```suggestion
    %%IMAGE%%
```

File: caddy/content.md [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+...
+$ curl http://localhost/
+hello world
+```
+
+To override the default [`Caddyfile`](https://github.com/caddyserver/dist/blob/master/config/Caddyfile), you can mount a new one at `/etc/caddy/Caddyfile`:
+
+```console
+$ docker run -d -p 80:80 \
+    -v $(pwd)/Caddyfile:/etc/caddy/Caddyfile \
```
Comment (tianon) [inline caddy/content.md:42]:
```suggestion
    -v $PWD/Caddyfile:/etc/caddy/Caddyfile \
```

File: caddy/content.md [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+Note that named volumes are persisted across container restarts and terminations, so if you move to a new image version, the same data and config directories can be re-used.
+
+### Basic Usage
+
+The default config file simply serves files from `/usr/share/caddy`, so if you want to serve `index.html` from the current working directory:
+
+```console
+$ echo "hello world" > index.html
+$ docker run -d -p 80:80 \
+    -v $(pwd)/index.html:/usr/share/caddy/index.html \
```
Comment (tianon) [inline caddy/content.md:30]:
```suggestion
    -v $PWD/index.html:/usr/share/caddy/index.html \
```

File: caddy/content.md [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+### Automatic TLS with the Caddy image
+
+The default `Caddyfile` only listens to port `80`, and does not set up automatic TLS. However, if you have a domain name for your site, and its A/AAAA DNS records are properly pointed to this machine's public IP, then you can use this command to simply serve a site over HTTPS:
+
+```console
+$ docker run -d -p 80:80 -p 443:443 \
+    -v /site:/usr/share/caddy \
+    -v caddy_data:/data \
+    -v caddy_config:/config \
+    caddy/caddy caddy file-server --domain example.com
```
Comment (tianon) [inline caddy/content.md:56]:
```suggestion
    %%IMAGE%% caddy file-server --domain example.com
```

File: caddy/content.md [hunk 9]
```diff
@@ -0,0 +1,110 @@
...
+```
+
+The key here is that Caddy is able to listen to ports `80` and `443`, both required for the ACME HTTP challenge.
+
+See [Caddy's docs](https://caddyserver.com/docs/automatic-https) for more information on automatic HTTPS support!
+
+### Building your own Caddy-based image
+
+Most users deploying production sites will not want to rely on mounting files into a container, but will instead base their own images on `caddy/caddy`:
```
Comment (tianon) [inline caddy/content.md:65]:
```suggestion
Most users deploying production sites will not want to rely on mounting files into a container, but will instead base their own images on `%%IMAGE%%`:
```

File: caddy/content.md [hunk 10]
```diff
@@ -0,0 +1,110 @@
...
+
+```Dockerfile
+# note: never use the :latest tag in a production site
+FROM caddy/caddy:v2.0.0
```
Comment (tianon) [inline caddy/content.md:69]:
```suggestion
FROM %%IMAGE%%:2.0.0
```

File: caddy/content.md [hunk 11]
```diff
@@ -0,0 +1,110 @@
...
+```
+
+#### Adding custom Caddy modules
+
+Caddy is extendable through the use of "modules". See https://caddyserver.com/docs/extending-caddy for full details.
+
+You can use the `:builder` image as a short-cut to building a new Caddy binary:
+
+```Dockerfile
+FROM caddy/caddy:v2.0.0-builder AS builder
```
Comment (tianon) [inline caddy/content.md:82]:
```suggestion
FROM %%IMAGE%%:2.0.0-builder AS builder
```

Comment (tianon) [inline caddy/content.md:82]:
```suggestion
FROM %%IMAGE%%:2.0.0-builder AS builder
```

File: caddy/content.md [hunk 12]
```diff
@@ -0,0 +1,110 @@
...
+
+RUN caddy-builder \
+    github.com/caddyserver/nginx-adapter \
+    github.com/hairyhenderson/caddy-teapot-module@v0.0.1
+
+FROM caddy/caddy:v2.0.0
```
Comment (tianon) [inline caddy/content.md:88]:
```suggestion
FROM %%IMAGE%%:2.0.0
```

File: caddy/content.md [hunk 13]
```diff
@@ -0,0 +1,110 @@
...
+
+COPY --from=builder /usr/bin/caddy /usr/bin/caddy
+```
+
+Note the second `FROM` instruction - this produces a much smaller image by simply overlaying the newly-built binary on top of the the regular `caddy/caddy` image.
```
Comment (tianon) [inline caddy/content.md:93]:
```suggestion
Note the second `FROM` instruction - this produces a much smaller image by simply overlaying the newly-built binary on top of the the regular `%%IMAGE%%` image.
```

**Comment by hairyhenderson** [comment]
Thanks @tianon - updated!

**Comment by hairyhenderson** [comment]
Yay! Thanks @yosifkit 😂

**Pull request merged**